examples/proof-of-concept/certificate-management/things/cmd/cert-management/main.go

```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright 2021 ForgeRock AS
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"crypto"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"os"
+	"regexp"
+
+	"encoding/pem"
+	"net/url"
+
+	"github.com/ForgeRock/iot-edge/examples/secrets"
+	"github.com/ForgeRock/iot-edge/v7/pkg/builder"
+	"github.com/ForgeRock/iot-edge/v7/pkg/callback"
+	"github.com/ForgeRock/iot-edge/v7/pkg/thing"
+	"gopkg.in/square/go-jose.v2"
+)
+
+const AmUrl = "https://iot.iam.example.com/am"
+
+type attributeHandler struct {
+	name  string
+	value string
+}
+
+type csrHandler struct {
+	thingID string
+	signer  crypto.PrivateKey
+}
+
+func handlerMatch(cb callback.Callback, outputName string) bool {
+	if cb.Type != "StringAttributeInputCallback" {
+		return false
+	}
+	name := ""
+	for _, e := range cb.Output {
+		if e.Name == "name" {
+			name = e.Value.(string)
+			break
+		}
+	}
+	if name != outputName {
+		return false
+	}
+	return true
+}
+
+func (a attributeHandler) Handle(cb callback.Callback) (bool, error) {
+	if !handlerMatch(cb, a.name) {
+		return false, nil
+	}
+	for i, e := range cb.Input {
+		if ok, _ := regexp.MatchString(`IDToken\d+`, e.Name); ok {
+			cb.Input[i].Value = a.value
+			fmt.Println("--> Providing", a.name, "=", a.value)
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (a csrHandler) Handle(cb callback.Callback) (bool, error) {
+	if !handlerMatch(cb, "thingProperties") {
+		return false, nil
+	}
+	for i, e := range cb.Input {
+		if ok, _ := regexp.MatchString(`IDToken\d+`, e.Name); ok {
+			csrPem := certificateSigningRequest(a.thingID, a.signer)
+			thingProperties, err := json.Marshal(struct {
+				CSR string `json:"csr"`
+			}{
+				CSR: csrPem,
+			})
+			if err != nil {
+				fmt.Println(err)
+				os.Exit(1)
```

Comment:
```
os.Exit(1)
```
to 
```
return false, err
```

If a callback returns an error, the authentication stops and returns the error back to the caller. `os.Exit` is used in other places in this example but best practice should be that callbacks to fail gracefully.